From 295e687e2b162d8e6c5925570f74d12ff4895aba Mon Sep 17 00:00:00 2001
From: Jets Yan <jets.yan@amlogic.com>
Date: Mon, 9 Nov 2015 19:28:37 +0800
Subject: [PATCH 5917/5965] PD#115018: hdmitx: audio compliance of hisense/tcl
 tv

    use 6144*4 for audio N of 480p.
    TV: hisense LED24K316
    TV: tcl L19P21

Change-Id: I9b7d2451cc5fc45a90be03855426bb4eaf00dba7
---
 arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c  |  17 ++-
 .../amlogic/hdmi/hdmi_tx/hdmi_tx_compliance.c | 132 +++++++++++++-----
 .../linux/amlogic/hdmi_tx/hdmi_tx_module.h    |   1 +
 3 files changed, 110 insertions(+), 40 deletions(-)

diff --git a/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c b/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
index d37c20cc0b68..11e61330207e 100755
--- a/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
+++ b/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
@@ -3627,17 +3627,20 @@ static int hdmitx_cntl_misc(hdmitx_dev_t* hdmitx_device, unsigned cmd, unsigned
             printk("TODO %s[%d]\n", __func__, __LINE__);
         break;
     case MISC_COMP_AUDIO:
-        if (argv == COMP_AUDIO_SET_N_6144x2) {
+		if (argv == COMP_AUDIO_SET_N_6144x2) {
 			if(hdmitx_device->cur_audio_param.type == CT_PCM){
 				if( (hdmitx_device->cur_VIC == HDMI_480p60) || (hdmitx_device->cur_VIC == HDMI_480p60_16x9) ||
-					(hdmitx_device->cur_VIC == HDMI_480i60) || (hdmitx_device->cur_VIC == HDMI_480i60_16x9) )
+				(hdmitx_device->cur_VIC == HDMI_480i60) || (hdmitx_device->cur_VIC == HDMI_480i60_16x9) )
 					hdmi_wr_reg(TX_SYS1_ACR_N_1, 0x30);
 			}
-        }
-        if (argv == COMP_AUDIO_SET_N_6144x3) {
-            hdmi_wr_reg(TX_SYS1_ACR_N_1, 0x48);
-            printk("TODO %s[%d]\n", __func__, __LINE__);
-        }
+		} else if (argv == COMP_AUDIO_SET_N_6144x3) {
+			hdmi_wr_reg(TX_SYS1_ACR_N_1, 0x48);
+			printk("TODO %s[%d]\n", __func__, __LINE__);
+		} else if (argv == COMP_AUDIO_SET_N_6144x4) {
+			if ((hdmitx_device->cur_VIC == HDMI_480p60) || (hdmitx_device->cur_VIC == HDMI_480p60_16x9) ||
+			(hdmitx_device->cur_VIC == HDMI_480i60) || (hdmitx_device->cur_VIC == HDMI_480i60_16x9))
+				hdmi_wr_reg(TX_SYS1_ACR_N_1, 0x60);
+		}
         break;
     default:
         hdmi_print(ERR, "misc: " "hdmitx: unknown cmd: 0x%x\n", cmd);
diff --git a/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx_compliance.c b/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx_compliance.c
index 84810d645c43..fee5b7985964 100755
--- a/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx_compliance.c
+++ b/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx_compliance.c
@@ -42,36 +42,101 @@
 #include <linux/amlogic/hdmi_tx/hdmi_tx_compliance.h>
 
 typedef struct {
-    char *ReceiverBrandName;
-    char *ReceiverProductName;
-    unsigned char blk0_chksum;
+	char *ReceiverBrandName;
+	char *ReceiverProductName;
+	unsigned char blk0_chksum;
+	unsigned char audio_N;
 }special_tv;
 
-static special_tv special_N_6144x2_tv_tab[]= {
-    /*SONY KDL-32R300B*/
-    {
-        .ReceiverBrandName="SNY",
-        .ReceiverProductName="SONY",
-        .blk0_chksum=0xf8,
-    },
-    /*TCL L19F3270B*/
-    {
-        .ReceiverBrandName="TCL",
-        .ReceiverProductName="MST6M16",
-        .blk0_chksum=0xa9,
-    },
-    /*Panasonic TH-32A400C*/
-    {
-        .ReceiverBrandName="MEI",
-        .ReceiverProductName="Panasonic-TV",
-        .blk0_chksum=0x28,
-    },
-    /*32PFL3390/T3*/
-    {
-        .ReceiverBrandName="PHL",
-        .ReceiverProductName="32PFL3390/T3",
-        .blk0_chksum=0x84,
-    },
+static special_tv special_audio_N_tv_tab[]= {
+	/*SONY KDL-32R300B*/
+	{
+		.ReceiverBrandName="SNY",
+		.ReceiverProductName="SONY",
+		.blk0_chksum=0xf8,
+		.audio_N=COMP_AUDIO_SET_N_6144x2,
+	},
+	/*TCL L19F3270B*/
+	{
+		.ReceiverBrandName="TCL",
+		.ReceiverProductName="MST6M16",
+		.blk0_chksum=0xa9,
+		.audio_N=COMP_AUDIO_SET_N_6144x2,
+	},
+	/*Panasonic TH-32A400C*/
+	{
+		.ReceiverBrandName="MEI",
+		.ReceiverProductName="Panasonic-TV",
+		.blk0_chksum=0x28,
+		.audio_N=COMP_AUDIO_SET_N_6144x2,
+	},
+	/*32PFL3390/T3*/
+	{
+		.ReceiverBrandName="PHL",
+		.ReceiverProductName="32PFL3390/T3",
+		.blk0_chksum=0x84,
+		.audio_N=COMP_AUDIO_SET_N_6144x2,
+	},
+
+	{
+/*
+TV: HISENSE LED24K316
+
+EDID block:
+00 ff ff ff ff ff ff 00 20 a3 29 00 01 00 00 00
+23 12 01 03 80 73 41 78 0a f3 30 a7 54 42 aa 26
+0f 50 54 25 c8 00 31 59 45 59 61 4f 01 01 01 01
+01 01 01 01 01 01 02 3a 80 18 71 38 2d 40 58 2c
+45 00 80 88 42 00 00 1e 8c 0a d0 8a 20 e0 2d 10
+10 3e 96 00 80 88 42 00 00 18 00 00 00 fc 00 48
+44 4d 49 20 20 20 20 0a 20 20 20 20 00 00 00 fd
+00 32 55 1f 45 0f 00 0a 20 20 20 20 20 20 01 00
+
+02 03 1d 72 23 09 07 07 47 90 05 04 03 11 14 1f
+83 01 00 00 68 03 0c 00 30 00 38 2d 00 f3 39 80
+18 71 38 2d 40 58 2c 45 00 c4 8e 21 00 00 1e d6
+09 80 a0 20 e0 2d 10 08 60 22 00 12 8e 21 08 08
+18 01 1d 00 bc 52 d0 1e 20 b8 28 55 40 c4 8e 21
+00 00 1e 8c 0a d0 90 20 40 31 20 0c 40 55 00 c4
+8e 21 00 00 18 00 00 00 00 00 00 00 00 00 00 00
+00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 85
+*/
+		.ReceiverBrandName="HEC",
+		.ReceiverProductName="HDMI",
+		.blk0_chksum=0x00,
+		.audio_N=COMP_AUDIO_SET_N_6144x4,
+	},
+
+	{
+/*
+TV: TCL L19P21
+
+EDID block:
+00 ff ff ff ff ff ff 00 50 6c 00 00 00 00 00 00
+05 14 01 03 80 3a 20 78 0a 0d c9 a0 57 47 98 27
+12 48 4c 23 08 00 81 80 01 01 01 01 01 01 01 01
+01 01 01 01 01 01 66 21 50 b0 51 00 1b 30 40 70
+36 00 44 40 21 00 00 18 0e 1f 00 80 51 00 1e 30
+40 80 37 00 10 4e 42 00 00 1e 00 00 00 fc 00 4d
+54 32 33 4c 0a 20 20 20 20 20 20 20 00 00 00 fd
+00 32 3c 0e 46 10 00 0a 20 20 20 20 20 20 01 85
+
+02 03 24 70 4f 14 13 12 11 16 15 05 04 03 02 07
+06 01 1f 10 23 09 07 07 83 01 00 00 67 03 0c 00
+10 00 b8 2d 01 1d 80 d0 72 1c 16 20 10 2c 25 80
+44 40 21 00 00 9e 01 1d 80 18 71 1c 16 20 58 2c
+25 00 44 40 21 00 00 98 01 1d 00 bc 52 d0 1e 20
+b8 28 55 40 44 40 21 00 00 1e 01 1d 00 72 51 d0
+1e 20 6e 28 55 00 44 40 21 00 00 1e 00 00 00 00
+00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 6b
+
+*/
+		.ReceiverBrandName="TCL",
+		.ReceiverProductName="MT23L",
+		.blk0_chksum=0x85,
+		.audio_N=COMP_AUDIO_SET_N_6144x4,
+	}
+
 };
 
 static special_tv samsung_future_tv_tab[]= {
@@ -156,11 +221,12 @@ void hdmitx_special_handler_video(hdmitx_dev_t* hdev)
  */
 void hdmitx_special_handler_audio(hdmitx_dev_t* hdev)
 {
-    int i = 0;
-    for(i = 0; i < ARRAY_SIZE(special_N_6144x2_tv_tab); i++) {
-        if (recoginze_tv(hdev, special_N_6144x2_tv_tab[i].ReceiverBrandName, special_N_6144x2_tv_tab[i].ReceiverProductName, special_N_6144x2_tv_tab[i].blk0_chksum))
-			hdev->HWOp.CntlMisc(hdev, MISC_COMP_AUDIO, COMP_AUDIO_SET_N_6144x2);
-    }
+	int i = 0;
+
+	for (i = 0; i < ARRAY_SIZE(special_audio_N_tv_tab); i++) {
+		if (recoginze_tv(hdev, special_audio_N_tv_tab[i].ReceiverBrandName, special_audio_N_tv_tab[i].ReceiverProductName, special_audio_N_tv_tab[i].blk0_chksum))
+			hdev->HWOp.CntlMisc(hdev, MISC_COMP_AUDIO, special_audio_N_tv_tab[i].audio_N);
+	}
 
 #if 0
 	// TODO
diff --git a/include/linux/amlogic/hdmi_tx/hdmi_tx_module.h b/include/linux/amlogic/hdmi_tx/hdmi_tx_module.h
index 5e88aac67f70..5c73194ac5a8 100644
--- a/include/linux/amlogic/hdmi_tx/hdmi_tx_module.h
+++ b/include/linux/amlogic/hdmi_tx/hdmi_tx_module.h
@@ -233,6 +233,7 @@ typedef struct hdmi_tx_dev_s {
 #define MISC_COMP_AUDIO         (CMD_MISC_OFFSET + 0x09)
     #define COMP_AUDIO_SET_N_6144x2          0x1
     #define COMP_AUDIO_SET_N_6144x3          0x2
+	#define COMP_AUDIO_SET_N_6144x4          0x3
 
 /***********************************************************************
  *                          Get State //GetState
-- 
2.19.0

